### PR TITLE
LPD-75909 Limit stored FragmentEntryVersion rows per fragment entry

### PR DESCRIPTION
**Pull request by georgel-pop-lr**
https://liferay.atlassian.net/browse/LPD-75909

Previous pull: https://github.com/brianchandotcom/liferay-portal/pull/177617

## What Is Being Fixed

Each edit to a fragment entry appends a new FragmentEntryVersion row, and nothing ever removes the old ones. A frequently edited fragment accumulates an unbounded number of version rows, growing the table without limit. Versions beyond the most recent serve no functional purpose.

## How It Is Being Fixed

A per-company configuration, FragmentEntryVersionConfiguration.maximumVersionsPerEntry, caps how many versions are retained per fragment entry. It defaults to 10, and 0 disables cleanup. FragmentEntryLocalService.cleanUpFragmentEntryVersions(long companyId) reads the configured cap and deletes every version beyond the newest count for each (ctCollectionId, fragmentEntryId) pair, while a no-arg cleanUpFragmentEntryVersions() orchestrates every company and isolates per-company failures. A daily scheduler job runs the cleanup, and a v3.0.3 upgrade prunes the existing backlog once.

The configuration mirrors the sibling ObjectEntryVersionConfiguration, which caps object entry versions the same way.

## PR Check

**pr-check: PASS** — tested on `8ab845ec6be27e5a9a338977d324eb940186e611`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Per-Module Compile | PASS |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS |
| Structural Smoke | PASS |

### CODE REVIEW

**Comment by liferay-page-management** [comment]
Thank you for submitting this PR. 👏

Make sure it follows the [Echo Pull Request Template](https://gist.github.com/liferay-echo/5a4455285ff1800f9da146816d85e67d#file-pull-request-template-md) if you haven't already done so. 

Learn more about it [here](https://liferay.atlassian.net/wiki/spaces/ENGECHO/pages/2027455014/Echo+Pull+Request+Etiquette).

File: modules/apps/fragment/fragment-api/src/main/java/com/liferay/fragment/configuration/FragmentEntryVersionConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	name = "fragment-entry-version-configuration-name"
+)
+public interface FragmentEntryVersionConfiguration {
+
+	@Meta.AD(
+		deflt = "10",
+		description = "maximum-versions-per-fragment-entry-description",
+		name = "maximum-versions-per-fragment-entry", required = false
+	)
+	public int maximumVersionsPerEntry();
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-api/src/main/java/com/liferay/fragment/configuration/FragmentEntryVersionConfiguration.java]: @georgel-pop-lr Should we make the method name consistent with the name using `maximumversionPerFragmentEntry`?

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	@Override
+	public UnsafeRunnable<Exception> getJobExecutorUnsafeRunnable() {
+		return _fragmentEntryLocalService::cleanUpFragmentEntryVersions;
+	}
+
+	@Override
+	public TriggerConfiguration getTriggerConfiguration() {
+		return TriggerConfiguration.createTriggerConfiguration(1, TimeUnit.DAY);
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java:82]: @georgel-pop-lr Maybe it would make sense to make this configurable to give control on when to run the cleanup

File: modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/change/tracking/test/FragmentEntryVersionCleanUpCTCollectionTest.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+	}
+
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/change/tracking/test/FragmentEntryVersionCleanUpCTCollectionTest.java:63]: @georgel-pop-lr I think we should add TestInfo annotation here too

File: modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/test/util/FragmentEntryVersionTestUtil.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Georgel Pop
+ */
+public class FragmentEntryVersionTestUtil {
+
+	public static final int MAX_VERSIONS_PER_ENTRY = 10;
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/test/util/FragmentEntryVersionTestUtil.java]: @georgel-pop-lr Should we name it `MAX_VERSIONS_PER_FRAGMENT_ENTRY`?

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java (hunk 2 above)
Comment (ruben-pulido) [reply modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java:82]: @georgel-pop-lr Here is an example: https://github.com/liferay/liferay-portal/blob/616d6d4f5f0cb92e4c769960e9c80b14a15eb23d/modules/apps/asset/asset-publisher-web/src/main/java/com/liferay/asset/publisher/web/internal/scheduler/CheckAssetEntrySchedulerJobConfiguration.java#L60-L81

File: modules/apps/fragment/fragment-api/src/main/java/com/liferay/fragment/configuration/FragmentEntryVersionConfiguration.java (hunk 1 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-api/src/main/java/com/liferay/fragment/configuration/FragmentEntryVersionConfiguration.java]: Done. Renamed the method to `maximumVersionsPerFragmentEntry()` so it matches the configuration key `maximum-versions-per-fragment-entry`.

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java (hunk 2 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java:82]: Done. Added a `cleanUpCronExpression` setting to the configuration. The scheduler reads it on activation and builds the trigger from it. If it is empty or invalid, it falls back to running once a day, so the default behavior is unchanged.

File: modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/change/tracking/test/FragmentEntryVersionCleanUpCTCollectionTest.java (hunk 3 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/change/tracking/test/FragmentEntryVersionCleanUpCTCollectionTest.java:63]: Done. Added `@TestInfo("LPD-75909")` to the test.

File: modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/test/util/FragmentEntryVersionTestUtil.java (hunk 4 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/test/util/FragmentEntryVersionTestUtil.java]: Done. Renamed the constant to `MAX_VERSIONS_PER_FRAGMENT_ENTRY`.

File: modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java (hunk 2 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-service/src/main/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfiguration.java:82]: Thanks for the example. I followed the same pattern: read the cron expression with `ConfigurableUtil` in `@Activate`, validate it through `TriggerFactory`, and fall back to the daily trigger when it is empty or invalid. Also added a unit test covering the empty, invalid, and valid cron cases.

File: modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		);
+
+		TriggerConfiguration triggerConfiguration =
+			(TriggerConfiguration)ReflectionTestUtil.invoke(
+				_cleanUpFragmentEntryVersionsSchedulerJobConfiguration,
+				"_getTriggerConfiguration", null, null);
+
+		Assert.assertNull(triggerConfiguration.getCronExpression());
+		Assert.assertEquals(1, triggerConfiguration.getInterval());
+		Assert.assertNotNull(triggerConfiguration.getTimeUnit());
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java]: @georgel-pop-lr Should we assert that this is `TimeUnit.DAY` instead of just checking for not null?

File: modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
```
Comment (ruben-pulido) [inline modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java]: @georgel-pop-lr Should we assert that this is `TimeUnit.DAY` instead of just checking for not null?

**Comment by ruben-pulido** [comment]
Hi @georgel-pop-lr 

The changes look good! I have left two minor improvements to the unit test. With that, the PR is ready from my side :)

File: modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java (hunk 5 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java]: Done, now asserting `TimeUnit.DAY` instead of just non-null.

File: modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java (hunk 6 above)
Comment (georgel-pop-lr) [reply modules/apps/fragment/fragment-service/src/test/java/com/liferay/fragment/internal/scheduler/CleanUpFragmentEntryVersionsSchedulerJobConfigurationTest.java]: Done, now asserting `TimeUnit.DAY` instead of just non-null.

**Comment by georgel-pop-lr** [comment]
@ruben-pulido Thanks a lot for the thorough review! I've addressed everything: renamed the config method to `maximumVersionsPerFragmentEntry()`, made the cleanup schedule configurable via `cleanUpCronExpression` (falling back to a daily trigger), added the `@TestInfo` annotation, renamed the constant to `MAX_VERSIONS_PER_FRAGMENT_ENTRY`, and the trigger tests now assert `TimeUnit.DAY`. All changes re-ran clean through pr-check. Thanks again!

**Comment by georgel-pop-lr** [comment]
**pr-check: PASS** — tested on `b08c4438099dcabfb6d03a884014a25365206f5c`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Per-Module Compile | PASS |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS |
| Structural Smoke | PASS |
| Java Unit Tests | PASS |

**Comment by georgel-pop-lr** [comment]
ci:forward

**Comment by georgel-pop-lr** [comment]
ci:forward
